gem/optimise.py

```diff
@@ -73,11 +129,234 @@ def filtered_replace_indices(node, self, subst):
+                return Zero(shape)
+        return Indexed(_select_expression(list(map(child, expressions)), index), multiindex)
+
+    if types <= {Literal, Zero, Failure}:
+        return partial_indexed(ListTensor(expressions), (index,))
+
+    if len(types) == 1:
+        cls, = types
+        if cls.__front__ or cls.__back__:
+            raise NotImplementedError
+        assert all(len(e.children) == len(expr.children) for e in expressions)
+        assert len(expr.children) > 0
+
+        return expr.reconstruct(*[_select_expression(nth_children, index)
+                                  for nth_children in zip(*[e.children
+                                                            for e in expressions])])
+
+    raise NotImplementedError
+
+
+def select_expression(expressions, index):
+    """Select an expression from a list of expressions with an index.
+    Semantically equivalent to
+
+        partial_indexed(ListTensor(expressions), (index,))
+
+    but has a much more optimised implementation.
+
+    :arg expressions: a list of expressions of the same shape
+    :arg index: an index (free, fixed or variable)
+    :returns: an expression of the same shape as the given expressions
+    """
+    # Check arguments
+    shape = expressions[0].shape
+    assert all(e.shape == shape for e in expressions)
+
+    # Sanitise input expressions
+    alpha = tuple(Index() for s in shape)
+    exprs = remove_componenttensors([Indexed(e, alpha) for e in expressions])
+
+    # Factor the expressions recursively and convert result
+    selected = _select_expression(exprs, index)
+    return ComponentTensor(selected, alpha)
+
+
+def delta_elimination(sum_indices, factors):
+    """IndexSum-Delta cancellation.
+
+    :arg sum_indices: free indices for contractions
+    :arg factors: product factors
+    :returns: optimised (sum_indices, factors)
+    """
+    sum_indices = list(sum_indices)  # copy for modification
+
+    delta_queue = [(f, index)
+                   for f in factors if isinstance(f, Delta)
+                   for index in (f.i, f.j) if index in sum_indices]
+    while delta_queue:
+        delta, from_ = delta_queue[0]
+        to_, = list({delta.i, delta.j} - {from_})
+
+        sum_indices.remove(from_)
+
+        mapper = MemoizerArg(filtered_replace_indices)
+        factors = [mapper(e, ((from_, to_),)) for e in factors]
+
+        delta_queue = [(f, index)
+                       for f in factors if isinstance(f, Delta)
+                       for index in (f.i, f.j) if index in sum_indices]
+
+    # Drop ones
+    return sum_indices, [e for e in factors if e != one]
+
+
+def sum_factorise(sum_indices, factors):
+    """Optimise a tensor product throw sum factorisation.
+
+    :arg sum_indices: free indices for contractions
+    :arg factors: product factors
+    :returns: optimised GEM expression
+    """
+    if len(sum_indices) > 5:
+        raise NotImplementedError("Too many indices for sum factorisation!")
```

Comment:
doi:10.1109/18.910572 does something like that, but that approach only yields an easy optimum if the factor graph is cycle-free which isn't the case for us. Cycle-breaking techniques can produce correct factorisations in this case, with no guarantee about optimality, so it's not really helpful for us.